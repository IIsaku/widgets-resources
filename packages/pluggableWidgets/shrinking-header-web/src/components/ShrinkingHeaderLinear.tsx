import { createElement, CSSProperties, ReactElement, ReactNode, useEffect, useState } from "react";
import classNames from "classnames";

import "../ui/ShrinkingHeader.scss";

export interface ShrinkingHeaderLinearProps {
    className?: string;
    style?: CSSProperties;
    tabIndex?: number;
    content?: ReactNode;
    initHeight: number;
    shrunkHeight: number;
}

export function ShrinkingHeaderLinear(props: ShrinkingHeaderLinearProps): ReactElement {
    const { className, style, tabIndex, content, initHeight, shrunkHeight } = props;

    const [headerHeight, setHeaderHeight] = useState<number>();

    const actualClassName = classNames("widget-shrinking-header", "widget-shrinking-header-linear", className);

    useEffect(() => {
        const updateHeaderHeight = function (): void {
            let headerHeight;

            if (window.scrollY > initHeight - shrunkHeight) {
                headerHeight = initHeight - (initHeight - shrunkHeight);
            } else if (window.scrollY < 0) {
                headerHeight = initHeight;
            } else {
                headerHeight = initHeight - window.scrollY;
            }

            setHeaderHeight(headerHeight);
        };

        updateHeaderHeight();

        document.addEventListener("scroll", updateHeaderHeight);

        return () => {
            document.removeEventListener("scroll", updateHeaderHeight);
        };
    }, [initHeight, shrunkHeight, setHeaderHeight]);

    return (
        <div className={actualClassName} style={{ ...style, height: initHeight }} tabIndex={tabIndex}>
            <header style={{ height: headerHeight }}>{content}</header>
        </div>
    );
}
