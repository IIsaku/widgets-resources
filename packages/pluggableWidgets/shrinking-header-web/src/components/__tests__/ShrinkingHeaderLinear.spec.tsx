import { createElement } from "react";
import { act } from "react-dom/test-utils";
import { mount } from "enzyme";

import { ShrinkingHeaderLinear, ShrinkingHeaderLinearProps } from "../ShrinkingHeaderLinear";
import { mockResizeObserver } from "../../utils/__tests__/ResizeObserverMock";

let lastMockedResizeObserver: ResizeObserver & { notifyChange: () => void };

mockResizeObserver(act, _lastMockedResizeObserver => {
    lastMockedResizeObserver = _lastMockedResizeObserver;
});

let documentEventListeners: Array<() => void>;

document.addEventListener = jest.fn().mockImplementation((_type: string, callback: () => void) => {
    documentEventListeners.push(callback);
});

describe("ShrinkingHeaderLinear", () => {
    let defaultShrinkingHeaderProps: ShrinkingHeaderLinearProps;

    beforeEach(() => {
        defaultShrinkingHeaderProps = {
            rootElementRef: jest.fn(),
            name: "name",
            className: "class-name",
            style: { color: "green" },
            tabIndex: 4,
            content: <h1>Header</h1>,
            initHeight: 100,
            shrunkHeight: 25
        };

        documentEventListeners = [];
    });

    it("renders correctly", () => {
        const shrinkingHeaderLinear = mount(<ShrinkingHeaderLinear {...defaultShrinkingHeaderProps} />);

        Object.defineProperty(shrinkingHeaderLinear.find("header").getDOMNode() as HTMLDivElement, "offsetHeight", {
            value: 100,
            configurable: true
        });
        lastMockedResizeObserver.notifyChange();
        shrinkingHeaderLinear.update();

        expect(shrinkingHeaderLinear).toMatchSnapshot();
        expect(defaultShrinkingHeaderProps.rootElementRef).toHaveBeenCalledTimes(1);
    });

    it("updates the header & wrapping div height when the page gets scrolled", () => {
        const shrinkingHeaderLinear = mount(<ShrinkingHeaderLinear {...defaultShrinkingHeaderProps} />);

        Object.defineProperty(shrinkingHeaderLinear.find("header").getDOMNode() as HTMLDivElement, "offsetHeight", {
            value: 100,
            configurable: true
        });
        lastMockedResizeObserver.notifyChange();
        shrinkingHeaderLinear.update();
        expect(documentEventListeners).toHaveLength(1);

        Object.defineProperty(window, "scrollY", { value: 25 });
        act(documentEventListeners[0]);
        Object.defineProperty(shrinkingHeaderLinear.find("header").getDOMNode() as HTMLDivElement, "offsetHeight", {
            value: 75,
            configurable: true
        });
        lastMockedResizeObserver.notifyChange();
        shrinkingHeaderLinear.update();
        expect(shrinkingHeaderLinear).toMatchSnapshot();
    });

    it("maintains the shrunk height as the header height when the page gets scrolled", () => {
        const shrinkingHeaderLinear = mount(<ShrinkingHeaderLinear {...defaultShrinkingHeaderProps} />);

        Object.defineProperty(window, "scrollY", { value: 125 });
        act(documentEventListeners[0]);
        shrinkingHeaderLinear.update();
        expect(shrinkingHeaderLinear.find("header").prop("style")).toEqual({ height: 25 });
    });

    // == Unmount doesn't trigger the cleanup function of the useEffect. This could be caused by erroneous code in React 17 as with both Enzyme and React Testing Library the function isn't triggered, even when waiting for it.
    // it("unmounts correctly", () => {
    //     const spy = jest.spyOn(document, "removeEventListener");
    //
    //     const shrinkingHeaderLinear = mount(<ShrinkingHeaderLinear {...defaultShrinkingHeaderProps} />);
    //
    //     expect(documentEventListeners).toHaveLength(1);
    //
    //     shrinkingHeaderLinear.unmount();
    //     expect(spy).toHaveBeenCalledWith("scroll", documentEventListeners[0]);
    // });
});
