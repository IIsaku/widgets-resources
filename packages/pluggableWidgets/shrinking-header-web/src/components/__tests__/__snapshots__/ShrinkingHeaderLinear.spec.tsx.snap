// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ShrinkingHeaderLinear renders correctly 1`] = `
<ShrinkingHeaderLinear
  className="class-name"
  content={
    <h1>
      Header
    </h1>
  }
  initHeight={100}
  shrunkHeight={25}
  style={
    Object {
      "color": "green",
    }
  }
  tabIndex={4}
>
  <div
    className="widget-shrinking-header widget-shrinking-header-linear class-name"
    style={
      Object {
        "color": "green",
        "height": 100,
      }
    }
    tabIndex={4}
  >
    <header
      style={
        Object {
          "height": 100,
        }
      }
    >
      <h1>
        Header
      </h1>
    </header>
  </div>
</ShrinkingHeaderLinear>
`;

exports[`ShrinkingHeaderLinear updates the header & wrapping div height when the page gets scrolled 1`] = `
<ShrinkingHeaderLinear
  className="class-name"
  content={
    <h1>
      Header
    </h1>
  }
  initHeight={100}
  shrunkHeight={25}
  style={
    Object {
      "color": "green",
    }
  }
  tabIndex={4}
>
  <div
    className="widget-shrinking-header widget-shrinking-header-linear class-name"
    style={
      Object {
        "color": "green",
        "height": 100,
      }
    }
    tabIndex={4}
  >
    <header
      style={
        Object {
          "height": 75,
        }
      }
    >
      <h1>
        Header
      </h1>
    </header>
  </div>
</ShrinkingHeaderLinear>
`;
